File: features/fintech/saldo_details/src/main/java/com/tokopedia/saldodetails/activity/SaldoDepositActivity.java
package com.tokopedia.saldodetails.activity;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.TextView;

import com.airbnb.deeplinkdispatch.DeepLink;
import com.tokopedia.abstraction.base.view.activity.BaseSimpleActivity;
import com.tokopedia.abstraction.common.di.component.HasComponent;
import com.tokopedia.applink.ApplinkConst;
import com.tokopedia.applink.RouteManager;
import com.tokopedia.saldodetails.R;
import com.tokopedia.saldodetails.commom.analytics.SaldoDetailsConstants;
import com.tokopedia.saldodetails.di.SaldoDetailsComponent;
import com.tokopedia.saldodetails.di.SaldoDetailsComponentInstance;
import com.tokopedia.saldodetails.presenter.SaldoDetailsPresenter;
import com.tokopedia.saldodetails.view.fragment.SaldoDepositFragment;
import com.tokopedia.user.session.UserSession;

import javax.inject.Inject;

@DeepLink(ApplinkConst.DEPOSIT)
public class SaldoDepositActivity extends BaseSimpleActivity implements
        HasComponent<SaldoDetailsComponent> {

    private static final int REQUEST_CODE_LOGIN = 1001;
    private static final String TAG = "DEPOSIT_FRAGMENT";

    @Inject
    UserSession userSession;
    private boolean isSeller;

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == SaldoDetailsPresenter.REQUEST_WITHDRAW_CODE && resultCode == Activity.RESULT_OK) {
            if (getSupportFragmentManager().findFragmentByTag(TAG) == null) {
                finish();
            } else {
                ((SaldoDepositFragment) getSupportFragmentManager().findFragmentByTag(TAG)).refresh();
            }
        }
        if (requestCode == REQUEST_CODE_LOGIN) {
            if (resultCode == RESULT_OK) {
                inflateFragment();
            } else {
                finish();
            }
        }
    }

    private void initInjector() {
        SaldoDetailsComponentInstance.getComponent(getApplication()).inject(this);
    }

    @DeepLink(ApplinkConst.DEPOSIT)
    public static Intent createInstance(Context context) {
        return new Intent(context, SaldoDepositActivity.class);
    }

    @Override
    public SaldoDetailsComponent getComponent() {
        return SaldoDetailsComponentInstance.getComponent(getApplication());
    }

    @Override
    protected int getLayoutRes() {
        return R.layout.activity_saldo_deposit;
    }

    @Override
    protected Fragment getNewFragment() {

        if (userSession.isLoggedIn()) {
            return SaldoDepositFragment.createInstance(isSeller);
        } else {
            startActivityForResult(RouteManager.getIntent(this, ApplinkConst.LOGIN), REQUEST_CODE_LOGIN);
            return null;
        }

    }

    @Override
    protected void setupLayout(Bundle savedInstanceState) {
        super.setupLayout(savedInstanceState);
        initInjector();
        initializeView();
        setUpToolbar();
    }

    private void setUpToolbar() {
        toolbar = findViewById(R.id.toolbar);
        Drawable upArrow = ContextCompat.getDrawable(this, R.drawable.ic_action_back);
        if (upArrow != null) {
            upArrow.setColorFilter(ContextCompat.getColor(this, R.color.grey_700), PorterDuff.Mode.SRC_ATOP);
            toolbar.setNavigationIcon(upArrow);
        } else {
            toolbar.setNavigationIcon(R.drawable.ic_icon_back_black);
        }
        toolbar.setPadding(toolbar.getPaddingLeft(), toolbar.getPaddingTop(), 30, toolbar.getPaddingBottom());
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowTitleEnabled(true);
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setTitle(this.getTitle());
            updateTitle("");
        }

    }

    private void initializeView() {
        isSeller = userSession.hasShop() || userSession.isAffiliate();
        TextView saldoHelp = findViewById(R.id.toolbar_saldo_help);

        if (isSeller) {
            saldoHelp.setVisibility(View.VISIBLE);
            saldoHelp.setOnClickListener(v -> {
                RouteManager.route(this, String.format("%s?url=%s", ApplinkConst.WEBVIEW,
                        SaldoDetailsConstants.SALDO_HELP_URL));
            });
        } else {
            saldoHelp.setVisibility(View.GONE);
        }
    }

    @Override
    protected void setupStatusBar() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            Window window = getWindow();
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            window.setStatusBarColor(ContextCompat.getColor(this, R.color.white));
            window.getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR);
        }
    }

}

File: features/fintech/saldo_details/src/main/java/com/tokopedia/saldodetails/saldoDetail/SaldoDepositActivity.kt
package com.tokopedia.saldodetails.saldoDetail

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import android.widget.RelativeLayout
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.google.android.material.snackbar.Snackbar
import com.tokopedia.abstraction.base.view.activity.BaseSimpleActivity
import com.tokopedia.abstraction.common.di.component.HasComponent
import com.tokopedia.applink.ApplinkConst
import com.tokopedia.applink.RouteManager
import com.tokopedia.applink.UriUtil
import com.tokopedia.applink.internal.ApplinkConstInternalGlobal
import com.tokopedia.header.HeaderUnify
import com.tokopedia.iconunify.IconUnify
import com.tokopedia.iconunify.getIconUnifyDrawable
import com.tokopedia.remoteconfig.FirebaseRemoteConfigImpl
import com.tokopedia.saldodetails.commom.di.component.SaldoDetailsComponent
import com.tokopedia.saldodetails.commom.di.component.SaldoDetailsComponentInstance
import com.tokopedia.saldodetails.saldoDetail.SaldoDepositFragment.Companion.REQUEST_WITHDRAW_CODE
import com.tokopedia.saldodetails.saldoDetail.coachmark.SaldoCoachMarkListener
import com.tokopedia.unifycomponents.Toaster
import com.tokopedia.user.session.UserSession
import com.tokopedia.unifyprinciples.R as unifyprinciplesR
import javax.inject.Inject
import com.tokopedia.saldodetails.R as saldodetailsR

/**
 * For navigating to this class
 * [com.tokopedia.applink.internal.ApplinkConstInternalGlobal.SALDO_DEPOSIT]
 */


class SaldoDepositActivity : BaseSimpleActivity(), HasComponent<SaldoDetailsComponent>, SaldoCoachMarkListener {

    var saldoToolbar: HeaderUnify? = null

    @Inject
    lateinit var userSession: UserSession
    private var isSeller: Boolean = false
    private val saldoComponent by lazy(LazyThreadSafetyMode.NONE) { SaldoDetailsComponentInstance.getComponent(this) }
    private var defaultStatusBarColor: Int = 0

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        val unMaskedRequestCode = requestCode and 0x0000ffff
        if ((requestCode == REQUEST_WITHDRAW_CODE || unMaskedRequestCode == REQUEST_WITHDRAW_CODE)
                && resultCode == Activity.RESULT_OK) {
            if (supportFragmentManager.findFragmentByTag(TAG) == null) {
                finish()
            } else {
                (supportFragmentManager.findFragmentByTag(TAG) as SaldoDepositFragment).resetPage()
            }
        }
        if (requestCode == REQUEST_CODE_LOGIN) {
            if (resultCode == Activity.RESULT_OK) {
                inflateFragment()
            } else {
                finish()
            }
        }
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)

        intent?.let {
            executeParamAction(it)
        }
    }

    private fun initInjector() {
        saldoComponent.inject(this)
    }

    override fun getComponent(): SaldoDetailsComponent? {
        return saldoComponent
    }

    override fun getLayoutRes(): Int {
        return com.tokopedia.saldodetails.R.layout.activity_saldo_deposit
    }

    override fun getNewFragment(): Fragment? {
        return if (userSession.isLoggedIn) {
            SaldoDepositFragment.createInstance(isSeller)
        } else {
            startActivityForResult(RouteManager.getIntent(this, ApplinkConst.LOGIN), REQUEST_CODE_LOGIN)
            null
        }
    }

    override fun getParentViewResourceID() = com.tokopedia.saldodetails.R.id.saldo_deposit_parent_view

    override fun setupLayout(savedInstanceState: Bundle?) {
        super.setupLayout(savedInstanceState)
        initInjector()
        setUpToolbar()
        initializeView()
        defaultStatusBarColor = window.statusBarColor
        hideStatusBar()
    }

    private fun executeParamAction(intent: Intent) {
        val queryParam = UriUtil.uriQueryParamsToMap(intent.data.toString())
        val toasterText = queryParam[TOASTER]
        val shouldReload = queryParam.containsKey(RELOAD)
        if (toasterText?.isNotEmpty() == true) {
            val rootView = findViewById<RelativeLayout>(saldodetailsR.id.il_main_content)
            val type = if (queryParam[TOASTER_ERROR] == "true") Toaster.TYPE_ERROR else Toaster.TYPE_NORMAL
            Toaster.build(rootView, toasterText, Snackbar.LENGTH_SHORT, type).show()
        }

        if (shouldReload) {
            (supportFragmentManager.findFragmentByTag(TAG) as SaldoDepositFragment).resetPage()
        }
    }

    private fun setUpToolbar() {
        saldoToolbar = findViewById(com.tokopedia.saldodetails.R.id.saldo_deposit_toolbar)
        saldoToolbar?.apply {
            headerView?.setTextColor(ContextCompat.getColor(context, unifyprinciplesR.color.Unify_NN900))
            navigationIcon = getIconUnifyDrawable(context, IconUnify.ARROW_BACK, ContextCompat.getColor(context, unifyprinciplesR.color.Unify_NN900))
            setNavigationOnClickListener {
                onBackPressed()
            }
        }
    }

    private fun hideStatusBar() {
        this.window?.let { window ->
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
            window.statusBarColor = Color.TRANSPARENT
        }
    }

    fun showStatusBar() {
        this.window?.let { window ->
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
            window.statusBarColor = defaultStatusBarColor
        }
    }

    private fun initializeView() {
        isSeller = userSession.hasShop() || userSession.isAffiliate
    }

    override fun getTagFragment() = TAG
    override fun getScreenName() = null

    companion object {
        private const val FLAG_APP_SALDO_AUTO_WITHDRAWAL = "app_flag_saldo_auto_withdrawal_v2"
        private val REQUEST_CODE_LOGIN = 1001
        private val TAG = "DEPOSIT_FRAGMENT"
        private const val TOASTER = "toaster"
        private const val TOASTER_ERROR = "toaster_error"
        private const val RELOAD = "reload"
    }

    override fun startCoachMarkFlow(anchorView: View?) {
        if (supportFragmentManager.findFragmentByTag(TAG) == null) { } else {
            (supportFragmentManager.findFragmentByTag(TAG) as SaldoDepositFragment).startSaldoCoachMarkFlow(anchorView)
        }
    }
}
